File: crates/knot-core/resources/typst.py
"""
Knot - Python support for literate programming with Typst

This module provides functions to convert Python objects (DataFrames, plots)
to Typst-compatible output via side-channel communication.
"""

import os
import json
import hashlib
import pickle
import types
import importlib
import sys
import pydoc
import builtins
from pathlib import Path
from typing import Any, Optional


def _get_base_dir() -> Path:
    """Get cache directory from environment or temp."""
    cache_dir = os.environ.get('KNOT_CACHE_DIR')
    if cache_dir:
        cache_path = Path(cache_dir)
        cache_path.mkdir(parents=True, exist_ok=True)
        return cache_path

    import tempfile
    return Path(tempfile.gettempdir())


def _write_metadata(metadata: dict) -> bool:
    """Write metadata to side-channel file."""
    metadata_file = os.environ.get('KNOT_METADATA_FILE')
    if not metadata_file:
        return False

    filepath = Path(metadata_file)

    # Read existing metadata
    existing = []
    if filepath.exists():
        try:
            with open(filepath, 'r') as f:
                content = f.read().strip()
                if content:
                    existing = json.loads(content)
        except (json.JSONDecodeError, IOError):
            existing = []

    # Append new metadata
    existing.append(metadata)

    # Write back as JSON array
    with open(filepath, 'w') as f:
        json.dump(existing, f)

    return True


def typst(obj: Any, **kwargs) -> Any:
    """Convert Python objects to Typst representations."""
    try:
        import matplotlib.figure
        if isinstance(obj, matplotlib.figure.Figure):
            return _typst_matplotlib(obj, **kwargs)
    except ImportError:
        pass

    try:
        from plotnine import ggplot
        if isinstance(obj, ggplot):
            return _typst_plotnine(obj, **kwargs)
    except ImportError:
        pass

    try:
        import pandas as pd
        if isinstance(obj, pd.DataFrame):
            return _typst_dataframe(obj, **kwargs)
    except ImportError:
        pass

    print(obj)
    return obj


def current_plot():
    """Get the current matplotlib figure."""
    try:
        import matplotlib.pyplot as plt
    except ImportError:
        raise RuntimeError("matplotlib is required for current_plot().")

    fig = plt.gcf()
    if not fig.get_axes():
        raise ValueError("Current figure is empty.")
    return fig


def _typst_matplotlib(fig, width=None, height=None, dpi=None, format=None):
    import io
    import matplotlib.pyplot as plt

    width = width or float(os.environ.get('KNOT_FIG_WIDTH', '7'))
    height = height or float(os.environ.get('KNOT_FIG_HEIGHT', '5'))
    dpi = dpi or int(os.environ.get('KNOT_FIG_DPI', '300'))
    format = format or os.environ.get('KNOT_FIG_FORMAT', 'svg')

    fig.set_size_inches(width, height)
    buf = io.BytesIO()
    fig.savefig(buf, format='svg', bbox_inches='tight')
    fig_hash = hashlib.sha256(buf.getvalue()).hexdigest()[:16]

    filename = f"plot_{fig_hash}.{format}"
    filepath = _get_base_dir() / filename
    fig.savefig(filepath, format=format, dpi=dpi, bbox_inches='tight')

    metadata = {'type': 'plot', 'path': str(filepath.absolute()), 'format': format}
    if not _write_metadata(metadata):
        plt.show()
    return fig


def _typst_plotnine(gg, **kwargs):
    fig = gg.draw()
    _typst_matplotlib(fig, **kwargs)
    return gg


def _typst_dataframe(df, index=False, **kwargs):
    df_string = df.to_string()
    df_hash = hashlib.sha256(df_string.encode()).hexdigest()[:16]
    filename = f"dataframe_{df_hash}.csv"
    filepath = _get_base_dir() / filename
    df.to_csv(filepath, index=index)
    metadata = {'type': 'dataframe', 'path': str(filepath.absolute())}
    if not _write_metadata(metadata):
        print(df)
    return df

# --- Session Management ---

def save_session(path):
    """Saves the global session (from __main__) including modules."""
    import sys
    try:
        import __main__
        main_dict = __main__.__dict__
        
        state = {'__knot_modules__': {}}
        for k, v in list(main_dict.items()):
            if k.startswith('__') or k in ['save_session', 'load_session', 'typst', 'current_plot']:
                continue
            
            if isinstance(v, types.ModuleType):
                state['__knot_modules__'][k] = v.__name__
                continue
                
            try:
                pickle.dumps(v)
                state[k] = v
            except:
                pass
                
        with open(path, 'wb') as f:
            pickle.dump(state, f)
        return True
    except Exception as e:
        print(f"Python Error in save_session: {e}", file=sys.stderr)
        return False

def load_session(path):
    """Restores a session into __main__."""
    import sys
    try:
        if not os.path.exists(path):
            return False
            
        import __main__
        main_dict = __main__.__dict__
            
        with open(path, 'rb') as f:
            state = pickle.load(f)
            
        modules = state.pop('__knot_modules__', {})
        for alias, name in modules.items():
            try:
                main_dict[alias] = importlib.import_module(name)
            except:
                pass
                
        main_dict.update(state)
        return True
    except Exception as e:
        print(f"Python Error in load_session: {e}", file=sys.stderr)
        return False

# --- LSP Support ---

def get_hover(topic):
    """Documentation from __main__ context."""
    try:
        import __main__
        main_dict = __main__.__dict__
        
        obj = main_dict.get(topic)
        if obj is None:
            try:
                obj = eval(topic, main_dict)
            except:
                pass
        
        if obj is not None:
            doc = pydoc.getdoc(obj)
            if not doc:
                doc = pydoc.render_doc(obj, renderer=pydoc.plaintext)
            return doc
        else:
            return pydoc.render_doc(topic, renderer=pydoc.plaintext)
    except Exception as e:
        return f"No help found for '{topic}'"

def get_completions(token):
    """Completions from __main__ context."""
    try:
        import __main__
        main_dict = __main__.__dict__
        
        if '.' in token:
            parts = token.split('.')
            base_name = parts[0]
            prefix = parts[-1]
            
            obj = main_dict.get(base_name)
            if obj is not None:
                for part in parts[1:-1]:
                    obj = getattr(obj, part, None)
                    if obj is None: break
            
            if obj is not None:
                return "\n".join([attr for attr in dir(obj) if attr.startswith(prefix) and not attr.startswith('_')])
        else:
            candidates = list(main_dict.keys()) + dir(builtins)
            return "\n".join([c for c in candidates if c.startswith(token) and not c.startswith('_')])
    except:
        return ""
    return ""

File: crates/knot-core/resources/typst.R
# Knot - R support for literate programming with Typst

# --- Session Management ---

.knot_save_session <- function(path) {
  tryCatch({
    # Save objects
    save.image(file = path)
    
    # Save loaded packages
    packages_path <- sub("\\.RData$", "_packages.rds", path)
    saveRDS(.packages(), packages_path)
    
    TRUE
  }, error = function(e) {
    message(sprintf("Error saving R session: %s", e$message))
    FALSE
  })
}

.knot_load_session <- function(path) {
  tryCatch({
    if (!file.exists(path)) return(FALSE)
    
    # Restore packages first
    packages_path <- sub("\\.RData$", "_packages.rds", path)
    if (file.exists(packages_path)) {
      pkgs <- readRDS(packages_path)
      # Suppress package startup messages
      invisible(lapply(pkgs, function(p) {
        tryCatch(library(p, character.only = TRUE), error = function(e) {})
      }))
    }
    
    # Restore objects into GlobalEnv
    load(file = path, envir = .GlobalEnv)
    
    TRUE
  }, error = function(e) {
    message(sprintf("Error loading R session: %s", e$message))
    FALSE
  })
}

# --- Constant Objects (Caching) ---

.knot_hash_object <- function(obj_name) {
  if (!requireNamespace("digest", quietly = TRUE)) {
    stop("Package 'digest' is required. Install with: install.packages('digest')")
  }
  
  if (!exists(obj_name, envir = .GlobalEnv)) {
    return("NONE")
  }
  
  obj <- get(obj_name, envir = .GlobalEnv)
  digest::digest(obj, algo = "xxhash64")
}

.knot_save_constant <- function(obj_name, path) {
  if (exists(obj_name, envir = .GlobalEnv)) {
    saveRDS(get(obj_name, envir = .GlobalEnv), file = path)
    TRUE
  } else {
    FALSE
  }
}

.knot_load_constant <- function(obj_name, path) {
  if (file.exists(path)) {
    assign(obj_name, readRDS(path), envir = .GlobalEnv)
    TRUE
  } else {
    FALSE
  }
}

# --- Output Formatting (Typst) ---

typst <- function(obj) {
  # Generic S3 method could be implemented here
  # For now, we handle basic types
  
  if (inherits(obj, "ggplot")) {
    return(.knot_save_plot(obj))
  } else if (is.data.frame(obj)) {
    return(.knot_save_dataframe(obj))
  }
  
  # Fallback
  print(obj)
  invisible(obj)
}

.knot_get_base_dir <- function() {
  dir <- Sys.getenv("KNOT_CACHE_DIR")
  if (dir == "") {
    dir <- tempdir()
  }
  dir.create(dir, recursive = TRUE, showWarnings = FALSE)
  dir
}

.knot_write_metadata <- function(metadata) {
  meta_file <- Sys.getenv("KNOT_METADATA_FILE")
  if (meta_file == "") return(FALSE)
  
  existing <- list()
  if (file.exists(meta_file)) {
    tryCatch({
      existing <- jsonlite::fromJSON(meta_file)
    }, error = function(e) {})
  }
  
  # Append new metadata (jsonlite handling of single object vs list is tricky, simplification here)
  # Ideally we append to the list
  combined <- c(existing, list(metadata))
  jsonlite::write_json(combined, meta_file, auto_unbox = TRUE)
  TRUE
}

.knot_save_plot <- function(plot_obj) {
  if (!requireNamespace("ggplot2", quietly = TRUE)) return(plot_obj)
  
  width <- as.numeric(Sys.getenv("KNOT_FIG_WIDTH", "7"))
  height <- as.numeric(Sys.getenv("KNOT_FIG_HEIGHT", "5"))
  dpi <- as.integer(Sys.getenv("KNOT_FIG_DPI", "300"))
  format <- Sys.getenv("KNOT_FIG_FORMAT", "svg")
  
  # Create stable hash
  # This is simplified; in production we might want to hash the object content
  # For now we use a random ID or rely on the chunk hash provided by Rust?
  # Rust provides execution context, but here we are inside R.
  # Let's use a random hash for the filename to match Python implementation style
  hash <- digest::digest(plot_obj, algo = "xxhash64")
  filename <- sprintf("plot_%s.%s", hash, format)
  filepath <- file.path(.knot_get_base_dir(), filename)
  
  ggplot2::ggsave(filepath, plot = plot_obj, width = width, height = height, dpi = dpi, device = format)
  
  metadata <- list(
    type = "plot",
    path = normalizePath(filepath, mustWork = FALSE),
    format = format
  )
  
  if (!.knot_write_metadata(metadata)) {
    print(plot_obj)
  }
  
  invisible(plot_obj)
}

.knot_save_dataframe <- function(df) {
  # Hash content
  hash <- digest::digest(df, algo = "xxhash64")
  filename <- sprintf("dataframe_%s.csv", hash)
  filepath <- file.path(.knot_get_base_dir(), filename)
  
  write.csv(df, filepath)
  
  metadata <- list(
    type = "dataframe",
    path = normalizePath(filepath, mustWork = FALSE)
  )
  
  if (!.knot_write_metadata(metadata)) {
    print(df)
  }
  
  invisible(df)
}

# --- LSP Support ---

.knot_get_hover <- function(topic) {
  tryCatch({
    h <- utils::help(topic)
    if (length(h) > 0) {
      tf <- tempfile()
      tools::Rd2txt(utils:::.getHelpFile(h), out = tf, options = list(underline_titles = FALSE))
      content <- readLines(tf)
      unlink(tf)
      paste(content, collapse = "\n")
    } else {
      ""
    }
  }, error = function(e) {
    ""
  })
}

.knot_get_completions <- function(token) {
  tryCatch({
    if (grepl("\\$", token)) {
      parts <- strsplit(token, "\\$")[[1]]
      obj_name <- parts[1]
      prefix <- if (length(parts) > 1) parts[2] else ""
      
      if (exists(obj_name, envir = .GlobalEnv)) {
        obj <- get(obj_name, envir = .GlobalEnv)
        if (is.list(obj) || is.data.frame(obj)) {
          n <- names(obj)
          matches <- n[startsWith(n, prefix)]
          paste(matches, collapse = "\n")
        } else {
          ""
        }
      } else {
        ""
      }
    } else {
      # Global completions
      matches <- utils::apropos(sprintf("^%s", token))
      paste(matches, collapse = "\n")
    }
  }, error = function(e) {
    ""
  })
}